tket2/src/portmatching/predicate.rs

```diff
@@ -0,0 +1,291 @@
+        let hugr = data.hugr();
+        match self {
+            Predicate::IsOpEqual(exp_match_op) => {
+                let Ok(PredicateArguments::Node(node)) = self.parse_arguments(args) else {
+                    panic!("ill-formed constraint");
+                };
+
+                let match_op: MatchOp = hugr.get_optype(node).into();
+                exp_match_op == &match_op
+            }
+            &Predicate::IsWireSource(exp_out_port) => {
+                let Ok(PredicateArguments::NodeAndWire(node, wire)) = self.parse_arguments(args)
+                else {
+                    panic!("ill-formed constraint");
+                };
+                wire.source() == exp_out_port && node == wire.node()
+            }
+            &Predicate::IsWireSink(exp_in_port) => {
+                let Ok(PredicateArguments::NodeAndWire(node, wire)) = self.parse_arguments(args)
+                else {
+                    panic!("ill-formed constraint");
+                };
+                if data.hugr().num_inputs(node) <= exp_in_port.index() {
+                    return false;
+                }
+                let Some((exp_out_node, exp_out_port)) =
+                    data.hugr().single_linked_output(node, exp_in_port)
+                else {
+                    return false;
+                };
+                exp_out_node == wire.node() && exp_out_port == wire.source()
+            }
+            Predicate::IsDistinctFrom { .. } => {
+                let Ok(PredicateArguments::ManyNodes(nodes)) = self.parse_arguments(args) else {
+                    panic!("ill-formed constraint");
+                };
+                let first_node = nodes[0];
+                nodes[1..].iter().all(|&n| n != first_node)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{portmatching::tests::circ_with_copy, Tk2Op};
+
+    use super::*;
+    use hugr::{ops::OpType, IncomingPort, Node, OutgoingPort, Port};
+    use itertools::Either;
+    use portmatching::EvaluatePredicate;
+    use rstest::rstest;
+
+    fn get_nodes_by_tk2op(circ: &Circuit, t2_op: Tk2Op) -> Vec<Node> {
+        let t2_op: OpType = t2_op.into();
+        circ.hugr()
+            .nodes()
+            .filter(|n| circ.hugr().get_optype(*n) == &t2_op)
+            .collect()
+    }
+
+    #[rstest]
+    fn test_check_node_op(circ_with_copy: Circuit) {
+        let pred = Predicate::IsOpEqual(Tk2Op::Rx.into());
+        let rx_ops = get_nodes_by_tk2op(&circ_with_copy, Tk2Op::Rx);
+        for rx in rx_ops {
+            assert!(pred.check(&[&HugrVariableValue::Node(rx)], &circ_with_copy,));
+        }
+    }
+
+    fn wire_pred(port: hugr::Port) -> Predicate {
+        match port.as_directed() {
+            Either::Left(in_port) => Predicate::IsWireSink(in_port),
+            Either::Right(out_port) => Predicate::IsWireSource(out_port),
+        }
+    }
+
+    #[rstest]
+    fn test_check_shared_edge(circ_with_copy: Circuit) {
```

Comment:
I don't think this covers all the predicates Are there more tests somewhere else and is there a CI check for coverage?

Reply:
Yes! There are more tests (that depend on other parts of the code, so I'll add them in an upcoming PR). The reason coverage tests don't run here is because this does not merge into `main`. I'll run coverage tests properly before the final merge.

Reply:
That said, you are right: I will add a test for other predicates